File: src/main/java/wav/demon/StatCraft/Commands/SC/SCArrowsShot.java
package wav.demon.StatCraft.Commands.SC;

import com.mysema.query.Tuple;
import com.mysema.query.sql.SQLQuery;
import com.mysema.query.types.path.NumberPath;
import org.bukkit.command.CommandSender;
import wav.demon.StatCraft.Commands.ResponseBuilder;
import wav.demon.StatCraft.Commands.SecondaryArgument;
import wav.demon.StatCraft.Magic.ProjectilesCode;
import wav.demon.StatCraft.Querydsl.Projectiles;
import wav.demon.StatCraft.Querydsl.QPlayers;
import wav.demon.StatCraft.Querydsl.QProjectiles;
import wav.demon.StatCraft.StatCraft;
import wav.demon.StatCraft.Util;

import java.util.LinkedList;
import java.util.List;

public class SCArrowsShot extends SCTemplate {

    public SCArrowsShot(StatCraft plugin) {
        super(plugin);
        this.plugin.getBaseCommand().registerCommand("arrowsshot", this);
    }

    @Override
    public boolean hasPermission(CommandSender sender, String[] args) {
        return sender.hasPermission("statcraft.user.arrowsshot");
    }

    @Override
    @SecondaryArgument({"distance", "farthest"})
    public String playerStatResponse(String name, List<String> args) {
        int total;
        int normal = 0;
        int flaming = 0;

        int normalDistance = 0;
        int flamingDistance = 0;

        int normalMaxThrow = 0;
        int flamingMaxThrow = 0;

        try {
            int id = plugin.getDatabaseManager().getPlayerId(name);
            if (id < 0)
                throw new Exception();

            SQLQuery query = plugin.getDatabaseManager().getNewQuery();
            if (query == null)
                return "Sorry, there seems to be an issue connecting to the database right now.";
            QProjectiles p = QProjectiles.projectiles;
            List<Projectiles> result = query.from(p).where(p.id.eq(id),
                p.type.eq(ProjectilesCode.NORMAL_ARROW.getCode()).or(p.type.eq(ProjectilesCode.FLAMING_ARROW.getCode()))
            ).list(p);

            for (Projectiles projectiles : result) {
                ProjectilesCode code = ProjectilesCode.fromCode(projectiles.getType());

                if (code == null)
                    continue;

                switch (code) {
                    case NORMAL_ARROW:
                        normal = projectiles.getAmount();
                        normalDistance = projectiles.getTotalDistance();
                        normalMaxThrow = projectiles.getMaxThrow();
                        break;
                    case FLAMING_ARROW:
                        flaming = projectiles.getAmount();
                        flamingDistance = projectiles.getTotalDistance();
                        flamingMaxThrow = projectiles.getMaxThrow();
                        break;
                }
            }

            total = normal + flaming;
        } catch (Exception e) {
            total = 0;
            normal = 0;
            flaming = 0;
            normalDistance = 0;
            flamingDistance = 0;
            normalMaxThrow = 0;
            flamingMaxThrow = 0;
        }

        String arg;
        if (args.size() > 0)
            arg = args.get(0);
        else
            arg = "";

        switch (arg) {
            case "distance":
                return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Arrows Shot Total Distance")
                    .addStat("Total", Util.distanceUnits(normalDistance + flamingDistance))
                    .addStat("Normal", Util.distanceUnits(normalDistance))
                    .addStat("Flaming", Util.distanceUnits(flamingDistance))
                    .toString();
            case "farthest":
                return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Arrows Shot Farthest Shot")
                    .addStat("Normal", Util.distanceUnits(normalMaxThrow))
                    .addStat("Flaming", Util.distanceUnits(flamingMaxThrow))
                    .toString();
            default:
                return new ResponseBuilder(plugin)
                    .setName(name)
                    .setStatName("Arrows Shot")
                    .addStat("Total", df.format(total))
                    .addStat("Normal", df.format(normal))
                    .addStat("Flaming", df.format(flaming))
                    .toString();
        }
    }

    @Override
    @SecondaryArgument({"distance", "farthest", "flaming"})
    public String serverStatListResponse(int num, List<String> args) {
        boolean distance = false;
        QProjectiles p = QProjectiles.projectiles;
        QPlayers pl = QPlayers.players;
        SQLQuery query = plugin.getDatabaseManager().getNewQuery();
        if (query == null)
            return "Sorry, there seems to be an issue connecting to the database right now.";

        NumberPath<Integer> path = null;
        String titlePrefix = "";
        String titlePostfix = "";
        String title;

        ProjectilesCode code = ProjectilesCode.NORMAL_ARROW;

        for (String arg : args) {
            switch (arg) {
                case "distance":
                    path = p.totalDistance;
                    titlePrefix = "Total Distance Fired - ";
                    distance = true;
                    break;
                case "farthest":
                    path = p.maxThrow;
                    titlePrefix = "Farthest ";
                    titlePostfix = " Shot";
                    distance = true;
                    break;
                case "flaming":
                    code = ProjectilesCode.FLAMING_ARROW;
                    break;
            }
        }

        if (path == null) {
            path = p.amount;
            titlePostfix = "s Shot";
        }

        title = titlePrefix + (code == ProjectilesCode.NORMAL_ARROW ? "Normal Arrow" : "Flaming Arrow") + titlePostfix;

        List<Tuple> list = query
            .from(p)
            .where(p.type.eq(code.getCode()))
            .leftJoin(pl)
            .on(p.id.eq(p.id))
            .groupBy(pl.name)
            .orderBy(path.desc())
            .limit(num)
            .list(pl.name, path);

        if (distance)
            return topListDistanceResponse(title, list);
        else
            return topListResponse(title, list);
    }

    @Override
    public List<String> onTabComplete(CommandSender sender, String[] args) {
        if (args[args.length -1].startsWith("-")) {
            boolean top = false;
            for (String s : args) {
                if (s.startsWith("-top"))
                    top = true;
            }
            List<String> result = new LinkedList<>();
            List<String> list = new LinkedList<>();
            list.add("-all");
            list.add("-distance");
            list.add("-farthest");
            if (top)
                list.add("-flaming");

            for (String s : list) {
                if (s.startsWith(args[args.length -1])) {
                    result.add(s);
                }
            }
            return result;
        } else {
            return super.onTabComplete(sender, args);
        }
    }
}

File: src/main/kotlin/com/demonwav/statcraft/commands/sc/SCArrowsShot.kt
/*
 * StatCraft Bukkit Plugin
 *
 * Copyright (c) 2016 Kyle Wood (DemonWav)
 * https://www.demonwav.com
 *
 * MIT License
 */

package com.demonwav.statcraft.commands.sc

import com.demonwav.statcraft.StatCraft
import com.demonwav.statcraft.Util
import com.demonwav.statcraft.commands.ResponseBuilderKt
import com.demonwav.statcraft.commands.SecondaryArgument
import com.demonwav.statcraft.magic.ProjectilesCode
import com.demonwav.statcraft.querydsl.QPlayers
import com.demonwav.statcraft.querydsl.QProjectiles
import com.mysema.query.types.path.NumberPath
import org.bukkit.command.CommandSender
import java.sql.Connection
import java.util.LinkedList

class SCArrowsShot(plugin: StatCraft) : SCTemplate(plugin) {

    init {
        plugin.baseCommand.registerCommand("arrowsshot", this)
    }

    override fun hasPermission(sender: CommandSender, args: Array<String>?) = sender.hasPermission("statcraft.user.arrowsshot")

    @SecondaryArgument("distance", "farthest")
    override fun playerStatResponse(name: String, args: List<String>, connection: Connection): String {
        val total: Int
        var normal = 0
        var flaming = 0

        var normalDistance = 0
        var flamingDistance = 0

        var normalMaxThrow = 0
        var flamingMaxThrow = 0

        try {
            val id = getId(name)
            if (id < 0) {
                throw Exception()
            }

            val query = plugin.databaseManager.getNewQuery(connection) ?:
                return "Sorry, there seems to be an issue connecting to the database right now"

            val p = QProjectiles.projectiles
            val result = query.from(p).where(
                p.id.eq(id),
                p.type.eq(ProjectilesCode.NORMAL_ARROW.code).or(p.type.eq(ProjectilesCode.FLAMING_ARROW.code))
            ).list(p)

            result.forEach { projectiles ->
                val code = ProjectilesCode.fromCode(projectiles.type) ?: return@forEach

                when (code) {
                    ProjectilesCode.NORMAL_ARROW -> {
                        normal += projectiles.amount
                        normalDistance += projectiles.totalDistance
                        normalMaxThrow = Math.max(normalMaxThrow, projectiles.maxThrow)
                    }
                    ProjectilesCode.FLAMING_ARROW -> {
                        flaming += projectiles.amount
                        flamingDistance += projectiles.totalDistance
                        flamingMaxThrow = Math.max(flamingMaxThrow, projectiles.maxThrow)
                    }
                    else -> {}
                }
            }

            total = normal + flaming
        } catch (e: Exception) {
            total = 0
            normal = 0
            flaming = 0
            normalDistance = 0
            flamingDistance = 0
            normalMaxThrow = 0
            flamingMaxThrow = 0
        }

        val arg: String
        if (args.size > 0) {
            arg = args[0]
        } else {
            arg = ""
        }

        when (arg) {
            "distance" -> {
                return ResponseBuilderKt.build(plugin) {
                    playerName { name }
                    statName { "Arrows Shot Total Distance" }
                    stats["Total"] = Util.distanceUnits(normalDistance + flamingDistance)
                    stats["Normal"] = Util.distanceUnits(normalDistance)
                    stats["Flaming"] = Util.distanceUnits(flamingDistance)
                }
            }
            "farthest" -> {
                return ResponseBuilderKt.build(plugin) {
                    playerName { name }
                    statName { "Farthest Arrows Shot" }
                    stats["Normal"] = Util.distanceUnits(normalMaxThrow)
                    stats["Flaming"] = Util.distanceUnits(flamingMaxThrow)
                }
            }
            else -> {
                return ResponseBuilderKt.build(plugin) {
                    playerName { name }
                    statName { "Arrows Shot" }
                    stats["Total"] = df.format(total)
                    stats["Normal"] = df.format(normal)
                    stats["Flaming"] = df.format(flaming)
                }
            }
        }
    }

    @SecondaryArgument("distance", "farthest", "flaming")
    override fun serverStatListResponse(num: Int, args: List<String>, connection: Connection): String {
        var distance = false
        val p = QProjectiles.projectiles
        val pl = QPlayers.players
        val query = plugin.databaseManager.getNewQuery(connection) ?:
            return "Sorry, there seems to be an issue connecting to the database right now."

        var path: NumberPath<Int>? = null
        var titlePrefix = ""
        var titlePostfix = ""
        val title: String

        var code = ProjectilesCode.NORMAL_ARROW

        for (arg in args) {
            when (arg) {
                "distance" -> {
                    path = p.totalDistance
                    titlePrefix = "Total Distance Fired - "
                    distance = true
                }
                "farthest" -> {
                    path = p.maxThrow
                    titlePrefix = "Farthest "
                    titlePostfix = " Shot"
                    distance = true
                }
                "flaming" -> code = ProjectilesCode.FLAMING_ARROW
            }
        }

        if (path == null) {
            path = p.amount
            titlePostfix = "s Shot"
        }

        title = titlePrefix + (if (code == ProjectilesCode.NORMAL_ARROW) "Normal Arrow" else "Flaming Arrow") + titlePostfix

        val list = query
            .from(p)
            .leftJoin(pl)
            .on(p.id.eq(pl.id))
            .where(p.type.eq(code.code))
            .groupBy(pl.name)
            .orderBy(path!!.desc())
            .limit(num.toLong())
            .list(pl.name, path.sum())

        if (distance) {
            return topListDistanceResponse(title, list)
        } else {
            return topListResponse(title, list)
        }
    }

    override fun onTabComplete(sender: CommandSender, args: Array<String>): List<String> {
        if (args[args.size - 1].startsWith("-")) {
            var top = false
            for (s in args) {
                if (s.startsWith("-top")) {
                    top = true
                }
            }
            val list = LinkedList<String>()
            list.add("-all")
            list.add("-distance")
            list.add("-farthest")
            if (top) {
                list.add("-flaming")
            }

            return list.filter({ s -> s.startsWith(args[args.size - 1]) }).toList()
        } else {
            return super.onTabComplete(sender, args)
        }
    }
}
